src/Database/LSMTree/Internal/Run/Index/Compact/Construction.hs

```diff
@@ -249,45 +250,54 @@ yield :: MCompactIndex s -> ST s (Maybe Chunk)
 yield MCompactIndex{..} = do
     pageNo <- readSTRef mciCurrentPageNumber
     if pageNo `mod` mciMaxChunkSize == 0 then do -- The current chunk is full
-      cPrimary <- VU.unsafeFreeze =<< readSTRef mciPrimary
-      (writeSTRef mciPrimary $!) =<< VUM.new mciMaxChunkSize
+      cPrimary <- VU.unsafeFreeze . NE.head =<< readSTRef mciPrimary
+      modifySTRef' mciPrimary . NE.cons =<< VUM.new mciMaxChunkSize
       modifySTRef' mciClashes . NE.cons =<< VUM.new mciMaxChunkSize
       modifySTRef' mciLargerThanPage . NE.cons =<< VUM.new mciMaxChunkSize
       pure $ Just (Chunk cPrimary)
     else -- the current chunk is not yet full
       pure Nothing
 
+newtype PrimaryArray = PrimaryArray (VU.Vector Word32)
+
 -- | Finalise incremental construction, yielding final chunks.
 --
 -- This function is unsafe, so do /not/ modify the 'MCompactIndex' after using
 -- 'unsafeEnd'.
 --
 -- INVARIANTS: see [construction invariants](#construction-invariants).
-unsafeEnd :: MCompactIndex s -> ST s (Maybe Chunk, FinalChunk)
+unsafeEnd :: MCompactIndex s -> ST s (Maybe Chunk, FinalChunk, PrimaryArray)
```

Comment:
I think it would be better UX to have this function return `ST s (Maybe Chunk, FinalChunk, CompactIndex)`. We can still keep the `fromChunks` function for tests, but otherwise the user doesn't have to apply  `fromChunks` themselves

Reply:
Ah, there's one slightly annoying point: To return `CompactIndex`, we need to import it. But the module where it is defined already imports the `Construction` module (and I think it makes sense to do so). So we might want to return a tuple of things, from which the compact index can then be constructed in its own module. There's a similar pattern in place with `unsafeFinalise` on `RunAcc`and `MRun`.
Alternatively we can of course say that `Construction` type modules should instead import the type they are constructing. I'll try that and see if there are any unpleasant consequences.

Reply:
Returning both final chunk and compact index seems quite redundant. I could also remove the `FinalChunk` type and have the following API for serialisation:

```haskell
headerBuilder :: BB.Builder
chunkBuilder :: Chunk -> BB.Builder
finalBuilder :: NumEntries -> CompactIndex -> BB.Builder

-- potentially even a non-incremental version:
builder :: NumEntries -> CompactIndex -> BB.Builder
builder numEntries index =
  headerBuilder <> chunkBuilder (Chunk (ciPrimary index)) <> finalBuilder index
```

Reply:
> Ah, there's one slightly annoying point: To return `CompactIndex`, we need to import it. But the module where it is defined already imports the `Construction` module (and I think it kind of makes sense to do so). So we might want to return a tuple of things, from which the compact index can then be constructed in its own module. There's a similar pattern in place with `unsafeFinalise` on `RunAcc`and `MRun`. Alternatively we can of course say that `Construction` type modules should instead import the type they are constructing. I'll try that and see if there are any unpleasant consequences.

Good point -- a `Compact.Types` or Compact.`Common` module could also be a solution for the import problem